app/services/chat_completion_a_p_i.rb

```diff
@@ -0,0 +1,154 @@
+class ChatCompletionAPI
+
+  # This is a lightweight wrapper around the OpenAI::Client gem that hides away a bunch of the complexity.
+  #
+  # ChatCompletionAPI.get_next_response("You are a comedian", ["Tell me a joke"], model: "gpt-4")
+  #
+
+  def self.api_key
+    Rails.env.test? ? "" : Current.user.openai_key
+  end
+
+  def self.get_next_response(system_message, chat_messages, params = {})
+    message_payload = [{
+      role: 'system',
+      content: system_message
+    }]
+
+    chat_messages.each_with_index do |msg, i|
+      role = (i % 2 == 0) ? 'user' : 'assistant'
+
+      message_payload << {
+        role: role,
+        content: msg
+      }
+    end
+
+    response = call_api({messages: message_payload}.merge(params)) # format: https://platform.openai.com/docs/api-reference/chat
+  end
+
+
+  private
+
+  def self.model_or_engine
+    "gpt-3.5-turbo-1106"
+  end
+
+  def self.url
+    "https://api.openai.com/v1/chat/completions"
+  end
+
+  def self.default_params
+    {
+      model: model_or_engine,
+      max_tokens: 500,                      # a sensible default
+      n: 1,
+      response_format: { "type": "text" },  # or json_object
+    }
+  end
+
+  def self.call_api(params)
+    params = default_params.deep_symbolize_keys.merge(params.symbolize_keys)
+    verify_params!(params)
+
+    client = OpenAI::Client.new(
+      access_token: api_key,
+      request_timeout: 240,
+    )
+
+    response = ""
+    retries = 0
+    keep_retrying = true
+
+    begin
+      verify_token_count!(params)
+      response = ""
+      finished_reason = nil
+
+      if Rails.env.test?
+        raw_response = client.chat(parameters: params)
+        response = JSON.parse(raw_response)&.dig("choices", 0, "message", "content")
```

Comment:
Since we're using a test mock, why do we need to special case the rails environment?

Reply:
I wanted to create a nice golden path for stubbing. Without this conditional, you either stub out the full method (`call_api`) and skip all the code execution, or you try to stub out `client.chat` which is complex.

I simplified this a bit more since you left this comment. It does still have the conceptual complexity of a branch, solely for the test environment. Maybe that's bad? But it has the added benefit of raising a really nice exception in the test environment because it's easy to not realize an API call is happening deep in a test.

Does it seem like a good solution, given all that?

Reply:
Actually, I know how to make this better! I'll clean it up a bit more.